### ISSUE TITLE
Ability to have layout of graph returned by viz.draw match that of another graph.

### ISSUE THREAD

**Issue by robertness** [py-why/pywhy-graphs#24]
**Is your feature request related to a problem? Please describe.**
Visual comparison of discovered graphs.  For example, the following is awkward because the layouts are different.

![image](https://user-images.githubusercontent.com/895554/199772942-cf09e03b-ffc1-4626-9299-e88538a80252.png)

**Describe the solution you'd like**
I'd like the layout of a graph returned by viz.draw to match the layout of another graph.

**Describe alternatives you've considered**
Something like this.
```
layout = get_dot_layout(G1)
draw(G2, layout=layout)
```
Setting the layout of a graph seems simple with the "pos" argument, but I'm not sure how to easily get the pos values when they are set by a layout engine like dot.

**Comment by siebert-julien** [comment]
Hi, I am pretty new to pywhy and I am not sure how pywhy-graphs manages the drawing internally, but this is a small example on how to do it with networkx:

```python
import networkx as nx

# create a first graph g1

nodes1 = [x for x in 'slteabxd']
edges1 = [
    ('a','t'),
    ('s','l'),
    ('s','b'),
    ('l','e'),
    ('l','x'),
    ('l','d'),
    ('t','d'),
    ('t','x'),
    ('t','e'),
    ('b','d'),
]

g1 = nx.Graph()
g1.add_nodes_from(nodes1)
g1.add_edges_from(edges1)

# create a second (very similar graph) g2

nodes2 = [x for x in 'slteabxdz']
edges2 = [
    ('a','t'),
    ('s','l'),
    ('s','b'),
    ('l','e'),
    ('l','x'),
    ('l','d'),
    ('t','d'),
    ('t','x'),
    ('t','e'),
    ('b','d'),
    ('l','z'),
    ('z','x')
]

g2 = nx.Graph()
g2.add_nodes_from(nodes2)
g2.add_edges_from(edges2)

# get the positions of the nodes of the first graph

pos1 = nx.spring_layout(g1)

# get the positions of the nodes of the second graph where some of the nodes are fixed (here I use the same nodes names in g1 and g2 so it is relatively straightforward

pos2 = nx.spring_layout(g2, pos=pos1, fixed=pos1.keys())

# plot and compare both graphs:

nx.draw(g1,with_labels=True,pos=pos1)

nx.draw(g2,with_labels=True,pos=pos2)
```

This is the results I obtained:

![image](https://user-images.githubusercontent.com/83209746/207438061-df314eef-8d90-4e15-ae0a-3d716f48087d.png)

![image](https://user-images.githubusercontent.com/83209746/207438132-1e3d8c96-0aa7-4cff-80f9-3968bae31ad1.png)


Like I said, I'm pretty new to pywhy, but if anyone has time to explain how the drawing is done behind the scene, I'd like to help.

**Comment by adam2392** [comment]
Hi @siebert-julien thanks for checking this out!

If you take a look at the source code, the `draw()` function mainly tries to use `pygraphviz` to draw graphs in a more involved fashion, so that way we can control the visualization and draw various different endpoints/edges (note this is not possible w/ networkx). 

To enable a similar layout, we would probably need the explicit positions of each node/edge from a resulting graph from `draw()` and then be able to feed that back into `draw()` where the second call visualizes a graph w/ the overlapping nodes in the same exact position. The edges might change though, which is fine.

So to begin, I would take a look at the source code and see how you can extract the resulting layout of a pygraphviz graph. Then examine how we would enable a fixed-layout on the nodes.

**Comment by siebert-julien** [comment]
Hi, I made a couple of checks.

1.  We can use networkx to generate the positions of the nodes for PAG, ADMG, and CDPAG classes (well, they all inherit from nx.MixedEdgeGraph so it's kind of obvious, but as I am not so familiar with the different classes, I checked it anyway).
2. We can use the "pos" attribute in Graphviz (see here https://www.graphviz.org/doc/info/attrs.html#d:pos) to enforce the nodes positions.

So my suggestion would be to add a "pos" argument to `draw()`. It would be of the same type as what is returned by the layout functions in networkx (see https://networkx.org/documentation/stable/reference/drawing.html#module-networkx.drawing.layout). It is usually a dictionary (str --> np.ndarray), see the following example:

```python
>>> nx.spring_layout(G)
{'x': array([ 1. , -0.34245137]), 'y': array([-0.2364603 , -0.99676684]), 'z': array([0.54454515, 0.82525722])}
```

Then when the `draw()` function creates the nodes in the DiGraph object we can add the pos attribute.

It would look like something like this:

````python
def draw(G: nx.MixedEdgeGraph, direction: Optional[str] = None, pos = None):
    # I'm not sure about the type pos should have
    ...
    for v in G.nodes:
        child = str(v)
        if pos:
            dot.node(child, shape=shape, height=".5", width=".5", pos=f"{pos[v][0]},{pos[v][1]}")
        else:
            dot.node(child, shape=shape, height=".5", width=".5")
    ...
```

@adam2392 what do you think? would that be a first start?

**Comment by adam2392** [comment]
Yeah that sounds good! I would have perhaps:

- draw take pos as an optional argument
- perhaps an example showing how to extract pos from an existing drawed graph and then feed back into draw again? Similar to what @robertness drew?

If you start a PR, I can review as we go.

**Comment by siebert-julien** [comment]
Hi,

Concerning the example, this is what I had in mind.
Note: It seems that graphviz uses the `pos` attribute  only to set the initial positions of the nodes, then the `render` function computes a new layout which makes it difficult for me at the moment to get exactly the same positions when the nodes and links are different (see below).

```python
import pywhy_graphs
import networkx as nx

from pywhy_graphs import CPDAG, PAG
from pywhy_graphs.viz import draw

# create some dummy graphs
G = nx.DiGraph([("x", "y"), ("z", "y"), ("z", "w"), ("xy", "x"), ("xy", "y")])
admg = pywhy_graphs.set_nodes_as_latent_confounders(G, ["xy"])
cpdag = CPDAG()
cpdag.add_edges_from(G.edges, cpdag.undirected_edge_name)
cpdag.orient_uncertain_edge("x", "y")
cpdag.orient_uncertain_edge("xy", "y")
cpdag.orient_uncertain_edge("z", "y")
pag = PAG()
pag.add_edges_from(G.edges, cpdag.undirected_edge_name)


# get the layout from networkx:
pos_G = nx.spring_layout(G)
# for the next graphs we fix the positions of some nodes using the `fixed` and `pos` arguments (note this only work with the `spring_layout` function...)
pos_admg = nx.spring_layout(admg, pos=pos_G, fixed=pos_G.keys())
pos_cpdag = nx.spring_layout(cpdag, pos=pos_G, fixed=pos_G.keys())
pos_pag = nx.spring_layout(pag, pos=pos_G, fixed=pos_G.keys())

# draw the graphs
dot_G = draw(G,pos=pos_G)
dot_admg = draw(admg,pos=pos_admg)
dot_cpdag = draw(cpdag,pos=pos_cpdag)
dot_pag = draw(pag,pos=pos_pag)

# render the graphs
dot_G.render(outfile="G.png", view=True, engine='neato')
dot_admg.render(outfile="admg.png", view=True, engine='neato')
dot_cpdag.render(outfile="cpdag.png", view=True, engine='neato')
dot_pag.render(outfile="pag.png", view=True, engine='neato')
```

These should give the following visualizations:
### G
![G_neato](https://user-images.githubusercontent.com/83209746/207818935-a0bf86e6-4be3-4e4b-ba4d-198064da0510.png)
### admg
![admg_neato](https://user-images.githubusercontent.com/83209746/207818940-0c26fa6d-2025-4b6d-81e7-67926bd3d312.png)
### cpdag
![cpdag_neato](https://user-images.githubusercontent.com/83209746/207818942-70ef1139-706c-4538-9e02-12ec9288926e.png)
### pag
![pag_neato](https://user-images.githubusercontent.com/83209746/207818937-7d9ff2b2-b4cb-4625-999a-03f74af70ff0.png)

As one can see the layout for the `admg` graph is not the same as the others. This comes from `graphviz`.
The layouts generated by `neworkx` are indeed the same. 

```python
>>> pos_G
{'x': array([-0.30843004,  0.21564894]), 'y': array([-0.44857214,  0.63371013]), 'z': array([ 0.28410451, -0.16603635]), 'w': array([ 1.        , -0.95981559]), 'xy': array([-0.52710233,  0.27649287])}
>>> pos_admg
{'x': array([-0.30843004,  0.21564894]), 'y': array([-0.44857214,  0.63371013]), 'z': array([ 0.28410451, -0.16603635]), 'w': array([ 1.        , -0.95981559])}
>>> pos_cpdag
{'x': array([-0.30843004,  0.21564894]), 'y': array([-0.44857214,  0.63371013]), 'z': array([ 0.28410451, -0.16603635]), 'w': array([ 1.        , -0.95981559]), 'xy': array([-0.52710233,  0.27649287])}
>>> pos_pag
{'x': array([-0.30843004,  0.21564894]), 'y': array([-0.44857214,  0.63371013]), 'z': array([ 0.28410451, -0.16603635]), 'w': array([ 1.        , -0.95981559]), 'xy': array([-0.52710233,  0.27649287])}
```

This is what you get if you do not specify the `engine` attribute in the `render()` function (it seems that the default engine is `dot` and that the `pos` attribute is skiped), see https://www.graphviz.org/doc/info/attrs.html#d:pos
### G
![G](https://user-images.githubusercontent.com/83209746/207816528-e01ab3b1-3158-4b7d-9890-1727f14644aa.png)
### admg
![admg](https://user-images.githubusercontent.com/83209746/207816526-e03b03fb-6bfb-49d2-92b0-d416c229162d.png)
### cpdag
![cpdag](https://user-images.githubusercontent.com/83209746/207816525-abdbaa15-1f52-4768-841f-2c10541a0d7a.png)
### pag
![pag](https://user-images.githubusercontent.com/83209746/207816520-1fa5269e-0fa9-4514-8f1d-c4b25976f998.png)

With the `fdp` engine, we get something similar to what `neato` provide
### G
![G_fdp](https://user-images.githubusercontent.com/83209746/207822732-bc32928f-5b7d-4bb3-88e6-febd340d7d16.png)
### admg
![admg_fdp](https://user-images.githubusercontent.com/83209746/207822724-d0c2d824-131d-4015-8811-8a82cd5e0754.png)
### cpdag
![cpdag_fdp](https://user-images.githubusercontent.com/83209746/207822730-7df173b6-eb64-4438-b201-8fc35eb20e03.png)
### pag
![pag_fdp](https://user-images.githubusercontent.com/83209746/207822734-732aac84-252f-4e92-adc0-e509dc46d50a.png)

In summary, for me it is still unclear how we can force the graphviz `render()` to fix the positions we give it. I am not sure if it is actually possible. I will have to investigate that point later (now I really have to get back to work).

**Comment by adam2392** [comment]
@siebert-julien can we use the `draw()` function instead? https://pygraphviz.github.io/documentation/stable/tutorial.html#layout-and-drawing

**Comment by siebert-julien** [comment]
@adam2392 I checked how to fix the positions, `graphviz` as a specific syntax for the `pos` attribute. One needs to add a `!` at the end of the position coordinates.  This is not indicated in the `pos` documentation directly (see https://graphviz.org/docs/attrs/pos/) but in the `point` documentation (see https://www.graphviz.org/docs/attr-types/point/).

The change in the `draw()` looks like this:
```python
    for v in G.nodes:
        child = str(v)
        if pos:
            dot.node(child, shape=shape, height=".5", width=".5", pos=f"{pos[v][0]},{pos[v][1]}!")
        else:
            dot.node(child, shape=shape, height=".5", width=".5")
``` 


Now the same example as above give the following images:

![admg](https://user-images.githubusercontent.com/83209746/207957589-761a6b6f-9785-44dc-8e9d-2a66492bd613.png)
![cpdag](https://user-images.githubusercontent.com/83209746/207957595-42c870a5-03c4-48c7-8903-f0181162ada2.png)
![G](https://user-images.githubusercontent.com/83209746/207957601-406b48e0-d120-4ddb-a708-cdf96067b2fb.png)
![pag](https://user-images.githubusercontent.com/83209746/207957606-647a0e7d-6182-4c5a-85e9-531ef9e10298.png)

As you can see, the positions are now fixed for all graphs.

I just need to find out how to scale the layout (by default the unit of the `pos` are inches). There are two possibilities here:
1. Scaling the layout directly after calling `nx.spring_layout()`. I have not found yet if there is a unit for the numbers computed, and what this unit could be.
2. `graphviz` allows for some graph attributes (either `scale` or `inputscale`), I am trying that now let's see how and if this works.

I am not sure which option is better. I would tend to let the user scale the layout instead of making `draw()` more complex. @adam2392 what's your opinion?

Scaling the layout can be easily done before calling `draw` for example by using the `k` attribute in the `nx.spring_layout()` function (there are many possibilities here).

Here is the previous example with `k=10`:

```python
import pywhy_graphs
import networkx as nx

from pywhy_graphs import CPDAG, PAG
from pywhy_graphs.viz import draw

# create some dummy graphs
G = nx.DiGraph([("x", "y"), ("z", "y"), ("z", "w"), ("xy", "x"), ("xy", "y")])
admg = pywhy_graphs.set_nodes_as_latent_confounders(G, ["xy"])
cpdag = CPDAG()
cpdag.add_edges_from(G.edges, cpdag.undirected_edge_name)
cpdag.orient_uncertain_edge("x", "y")
cpdag.orient_uncertain_edge("xy", "y")
cpdag.orient_uncertain_edge("z", "y")
pag = PAG()
pag.add_edges_from(G.edges, cpdag.undirected_edge_name)


# get the layout from networkx:
pos_G = nx.spring_layout(G, k=10)
# for the next graphs we fix the positions of some nodes using the `fixed` and `pos` arguments (note this only work with the `spring_layout` function...)
pos_admg = nx.spring_layout(admg, pos=pos_G, fixed=pos_G.keys())
pos_cpdag = nx.spring_layout(cpdag, pos=pos_G, fixed=pos_G.keys())
pos_pag = nx.spring_layout(pag, pos=pos_G, fixed=pos_G.keys())


# draw the graphs
dot_G = draw(G,pos=pos_G)
dot_admg = draw(admg,pos=pos_admg)
dot_cpdag = draw(cpdag,pos=pos_cpdag)
dot_pag = draw(pag,pos=pos_pag)

# render the graphs
dot_G.render(outfile="G.png", view=True, engine='neato')
dot_admg.render(outfile="admg.png", view=True, engine='neato')
dot_cpdag.render(outfile="cpdag.png", view=True, engine='neato')
dot_pag.render(outfile="pag.png", view=True, engine='neato')
```

Here are the results:

![admg](https://user-images.githubusercontent.com/83209746/207961513-d644d4d4-e567-4205-9d4b-ff581236a00a.png)
![cpdag](https://user-images.githubusercontent.com/83209746/207961517-7857d6a8-57fe-487f-a311-4e8f8a6d2ca8.png)
![G](https://user-images.githubusercontent.com/83209746/207961519-37b6aecb-fdde-4134-9098-8bfd2fb1fa60.png)
![pag](https://user-images.githubusercontent.com/83209746/207961520-4559e4e7-d776-4942-b5c7-6b5652e69b7b.png)


@robertness  @adam2392 if this looks like the solution you envisaged, I'll prepare a PR, including tests. What do you think?

**Comment by adam2392** [comment]
@siebert-julien this looks like a good starting proposal! Yeah feel free to open up a PR and then include an example in the `examples/` directory, which we can render in the documentation as well.

Thanks for the in-depth explanations :)
